Document how to use tenant-provided TLS certificates

philandstuff opened the pull request:
Now that we have per-namespace ingressgateways, it's possible for
tenants to provide their own TLS certificates.  This updates the
documentation to show how.

This also shows how to use a custom domain, provided the tenant
provides the certificate.

I tried as much as possible to show the details of exactly how to do
each step.  In particular, I wanted to show how to create the `Secret`
and to seal it, and I wanted to make it easy to do the right thing and
hard to accidentally commit an unsealed seacret.  Hence my slightly
funky bash pipeline to do it all in one line.

PR #599 is a separate approach which would allow tenants to use
cluster-provisioned certificates with custom domains.

philandstuff committed:
Document how to use tenant-provided TLS certificates

Now that we have per-namespace ingressgateways, it's possible for
tenants to provide their own TLS certificates.  This updates the
documentation to show how.

This also shows how to use a custom domain, provided the tenant
provides the certificate.

I tried as much as possible to show the details of exactly how to do
each step.  In particular, I wanted to show how to create the `Secret`
and to seal it, and I wanted to make it easy to do the right thing and
hard to accidentally commit an unsealed seacret.  Hence my slightly
funky bash pipeline to do it all in one line.

PR #599 is a separate approach which would allow tenants to use
cluster-provisioned certificates with custom domains.

blairboy362 commented on docs/gds-supported-platform/tls-certificates.md:
```diff
@@ -27,6 +32,76 @@ spec:
+      protocol: HTTPS
+    tls:
+      mode: SIMPLE
+      credentialName: my-custom-cert
```
```suggestion
      credentialName: my-custom-cert
      serverCertificate: sds
      privateKey: sds
```

blairboy362 replied:
See https://istio.io/docs/reference/config/networking/v1alpha3/gateway/#Server-TLSOptions
`serverCertificate` and `privateKey` are required when TLS mode is `SIMPLE` or `MUTUAL`. Populating them with "sds" seems to be the convention when using SDS (as we are), but I guess they can be anything.

philandstuff replied:
Huh. Well despite what the docs say I tried this as-is (ie without those fields) and it worked fine, so maybe "REQUIRED" doesn't really mean "REQUIRED".

That said, I don't mind adding these fields to the docs.

blairboy362 replied:
Interesting. I wonder if that got "fixed" between 1.1.4 and 1.1.12 then. I'm sure I tested that at some point and couldn't get it to work. :man_shrugging: I'm happy to go with either option in that case.

blairboy362 commented on docs/gds-supported-platform/tls-certificates.md:
```diff
@@ -27,6 +32,76 @@ spec:
+As the example shows, if you provide your own certificates, you can
+use any custom domain you choose; you do not have to use the cluster
+`govsvc.uk` domain.  To use a custom domain, create a CNAME record
+from your domain to the external load balancer for your namespace.
```
As a possible alternative to this...we may want to encourage an additional level of indirection: add a `.govsvc.uk` record and then have the custom domain CNAME to that. That way, if the `Service` that is responsible for the NLB gets recreated (we've seen that happen several times lately; granted more in Sandbox but it's happened twice in Verify) their "custom" CNAME will self-correct once `external-dns` does its job on the newly-created NLB.

philandstuff replied:
If we want to do this, we should either a) document how tenants can add a `.govsvc.uk` record, b) add a `.govsvc.uk` record ourselves.  I'm guessing you had one of these options in mind but I can't tell from your comment which one.  I think I'd prefer the second one.  How about `<namespace>-ingress.london.<cluster>.govsvc.uk`?  Or maybe plain `<namespace>.london.<cluster>.govsvc.uk`?

blairboy362 replied:
Is https://github.com/alphagov/gsp/blob/master/docs/gds-supported-platform/external-dns.md not enough?

And I hadn't thought about adding one by default...If we do, of your suggestions, `<namespace>.<cluster-domain>` gets my vote.

philandstuff replied:
I've pushed an update that registers a standard DNS name for each ingressgateway

philandstuff committed:
create a standard external DNS name for each ingressgateway

So that users can target it with CNAMEs when using a custom domain.

jonathanglassman commented on docs/gds-supported-platform/tls-certificates.md:
```diff
@@ -1,5 +1,10 @@
 # Public TLS Certificates
 
+All apps in GSP are protected by TLS.  There are two options for this:
+using cluster-provided certificates, or providing your own.
```
Does the user have to enable the relevant one of these (i.e. TLS isnt enabled until one of these options is enabled)? 

Or is one enabled by default and the user needs to enable the other in order for it to work?

jonathanglassman commented on docs/gds-supported-platform/tls-certificates.md:
```diff
@@ -27,6 +32,75 @@ spec:
+ - you are in a namespace called `my-namespace`
+ - your certificate and key are in files called `my-cert.pem` and
+   `my-cert.key`
+ - you want to create a `SealedSecret` called `my-custom-cert` in a
```
Line 39 states that you will need a `SealedSecret` resource containing the certificate and key. Yet this section assumes that "you want to create a `SealedSecret` called `my-custom-cert`". At this stage, do you have the `SealedSecret` or not?

philandstuff replied:
ah. line 39 means "to achieve this task, you will need to create a SealedSecret and a Gateway" and the rest of the section explains how to go about creating these things.  Maybe it could be worded better because I can see how it's confusing.

philandstuff committed:
content fixes

Co-authored-by: Jonathan Glassman <jonathan.glassman@digital.cabinet-office.gov.uk>

blairboy362 commented on docs/gds-supported-platform/tls-certificates.md:
```diff
@@ -1,5 +1,12 @@
 # Public TLS Certificates
 
+All apps in GSP are protected by TLS.  By default, TLS will use a
```
> All apps in GSP are protected by TLS.

If they're configured to...this is up to tenants...

> By default, TLS will use a cluster-provided certificate.

Again, if it's configured to do so...up to tenants to provide a `Certificate` resource

philandstuff replied:
> If they're configured to...this is up to tenants...

I mean, we only listen on port 443 and we upgrade 80 to 443.  I suppose tenants could serve plain http on port 443 but I don't think saying "this is up to tenants" would be very helpful in these docs.

> Again, if it's configured to do so...up to tenants to provide a Certificate resource

ah, I hadn't understood the nuances here. I'll update the text for this.

blairboy362 replied:
> I mean, we only listen on port 443 and we upgrade 80 to 443.

Again: if the tenant namespace configuration does that; it's entirely up to them. The ingressgateway, by default, will listen on 443 and 80 (among others) but they can add other specifications via `-cluster-config`. And redirecting from 80 to 443 is controlled by the `Gateway`, which is also tenant-controlled.

philandstuff replied:
I rewrote the opening paragraph, I hope it addresses your concerns?

blairboy362 replied:
👍 LGTM

philandstuff committed:
Rewrite opening paragraph

the cluster doesn't provide a cert "by default" - the tenant still has
to create a Certificate object.
